### DISCUSSION TITLE
Request: pg_stat_activity_count

### DISCUSSION THREAD

**Discussion by prime-minister-of-fun** [pgexporter/pgexporter#136] [General]
Hi,

I'm porting a very widely used PostgreSQL dashboard to use your exporter and I don't think your exporter has the pg_stat_activity_count from another very popular exporter.

Here's the query.
```
SELECT
				pg_database.datname,
				tmp.state,
				tmp2.usename,
				tmp2.application_name,
				COALESCE(count,0) as count,
				COALESCE(max_tx_duration,0) as max_tx_duration
			FROM
				(
				  VALUES ('active'),
				  		 ('idle'),
				  		 ('idle in transaction'),
				  		 ('idle in transaction (aborted)'),
				  		 ('fastpath function call'),
				  		 ('disabled')
				) AS tmp(state) CROSS JOIN pg_database
			LEFT JOIN
			(
				SELECT
					datname,
					state,
					usename,
					application_name,
					count(*) AS count,
					MAX(EXTRACT(EPOCH FROM now() - xact_start))::float AS max_tx_duration
				FROM pg_stat_activity GROUP BY datname,state,usename,application_name) AS tmp2
				ON tmp.state = tmp2.state AND pg_database.datname = tmp2.datname
			`,
		},
		{
			semver.MustParseRange("<9.2.0"),
			`
			SELECT
				datname,
				'unknown' AS state,
				usename,
				application_name,
				COALESCE(count(*),0) AS count,
				COALESCE(MAX(EXTRACT(EPOCH FROM now() - xact_start))::float,0) AS max_tx_duration
			FROM pg_stat_activity GROUP BY datname,usename,application_name;

```

Here's the link: https://github.com/prometheus-community/postgres_exporter/blob/a4ac0e674769805a9b282edb9a7791e64cd977e3/cmd/postgres_exporter/queries.go#L107

**Comment by jesperpedersen** [comment]
Would you like to write a patch for it ?

Most of the work would be in

```
src/include/internal.h
contrib/yaml/*
```

And, you don't have to worry about `< 9.2` since we only support PostgreSQL 12+ now officially.

Or, are you planning to use the custom YAML functionality for it ?

**Reply by prime-minister-of-fun** [reply]
I'm happy to give it a try, but, C is way out of my shell scripting wheelhouse.
It seems like I add it to internal.h in a similar pattern/way as the other queries.  Let me know if I am off target.

**Reply by jesperpedersen** [reply]
Nope, that would be correct - find a query, insert yours, then `git grep` for the identifiers to see if there are other places - like documentation and so on.

https://github.com/pgexporter/pgexporter/blob/main/doc/DEVELOPERS.md will help you

**Reply by jesperpedersen** [reply]
Start by creating an issue for it that gives an overall description and the query itself

**Reply by jesperpedersen** [reply]
And think about: is the query different between PostgreSQL 12, 13, 14, 15 and 16...

**Reply by jesperpedersen** [reply]
It should be a fun project for you in C :smiley:
